File: Python/textbook_track/chapter03/ba3l.py
# Construct a String Spelled by a Gapped Genome Path
import sys

from collections import namedtuple


class GappedPattern(namedtuple('GappedPattern', ['first', 'second'])):
    __slots__ = ()

    @property
    def prefix(self):
        return GappedPattern(self.first[:-1], self.second[:-1])

    @property
    def suffix(self):
        return GappedPattern(self.first[1:], self.second[1:])


def convert_to_intlist(line):
    return [int(elem) for elem in line.split()]


def read_gapped_patterns(lines):
    return [GappedPattern(*line.split('|')) for line in lines]


def calc_string_spelled_by_a_gapped_genome_path(paired_patterns, k, d):
    def calc_string_spelled_by_a_genome_path(k_mers):
        text = list(k_mers[0])
        for k_mer in k_mers[1:]:
            text.append(k_mer[-1])
        return ''.join(text)

    first_k_mers, second_kmers = zip(*paired_patterns)
    first_genome = calc_string_spelled_by_a_genome_path(first_k_mers)
    second_genome = calc_string_spelled_by_a_genome_path(second_kmers)
    if first_genome[k+d:] != second_genome[:-(k+d)]:
        return None
    return first_genome[:k+d] + second_genome


def main():
    data = sys.stdin.read().splitlines()
    k, d = convert_to_intlist(data[0])
    paired_patterns = read_gapped_patterns(data[1:])
    result = calc_string_spelled_by_a_gapped_genome_path(paired_patterns, k, d)
    print(result)


if __name__ == '__main__':
    main()

File: Scala/src/main/scala/TextbookTrack/Chapter03/BA3L.scala
package TextbookTrack.Chapter03

object BA3L {
  final case class GappedPattern(first: String, second: String) {
    def prefix: GappedPattern = GappedPattern(first.dropRight(1), second.dropRight(1))

    def suffix: GappedPattern = GappedPattern(first.drop(1), second.drop(1))
  }

  private def readParameters(line: String): (Int, Int) = line.split(" ").map(_.toInt).toList match {
    case List(k, d) => (k, d)
    case _ => throw new Exception("Unexpected input data format.")
  }

  private def readGappedPatterns(reader: Iterator[String]): List[GappedPattern] =
    reader.map{ line =>
      line.split("\\|").toList match {
        case List(first, second) => GappedPattern(first, second)
        case _ => throw new Exception("Unexpected input data format.")
      }
    }.toList

  def calcStringSpelledByAGappedGenomePath(pairedPatterns: List[GappedPattern], k: Int, d: Int): Option[String] = {
    def calcStringSpelledByAGenomePath(kMers: List[String]): String = kMers match {
      case Nil => ""
      case x :: xs => (x.toList ::: xs.map(_.last)).mkString
    }

    val firstGenome: String = calcStringSpelledByAGenomePath(pairedPatterns.map(_.first))
    val secondGenome: String = calcStringSpelledByAGenomePath(pairedPatterns.map(_.second))
    if (firstGenome.drop(k + d) != secondGenome.dropRight(k + d)) None
    else Some(firstGenome.take(k + d) + secondGenome)
  }

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val (k, d): (Int, Int) = readParameters(reader.next())
    val pairedPatterns: List[GappedPattern] = readGappedPatterns(reader)
    calcStringSpelledByAGappedGenomePath(pairedPatterns, k, d) match {
      case None => println("No string spelled by the gapped pattern")
      case Some(genome) => println(genome)
    }
  }
}
